### PR TITLE
Rjf/state model deserialize

### LINKED ISSUE #138: Update state model to not use hashmap for low number of entries
Right now the state model uses a hash map to store state features in the state model. We can update the state model to use a different, faster mechanism to store the state features for a low number of features. 

From @robfitzgerald:

totally. that's how [Scala implements `Map`. for Map instances with 4 or less elements](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/immutable/Map.scala#L259), they rely on a scan, and don't actually instantiate a hashmap until 5+. that could make sense, and perhaps we pull it off by turning StateModel into:

```rust
pub enum StateModel {
  OneFeature {
    name: String,
    entry: StateModelEntry
  },
  TwoFeatures { 
    name1: String,
    name2: String,
    entry1: StateModelEntry,
    entry2: StateModelEntry,
  },
  ...
  NFeatures(HashMap<String, StateModelEntry>),
}
```

### LINKED ISSUE #140: state model should deserialize from a JSON object, not array
StateModel is deserialized from a TOML array. as JSON, it appears as:

```json
[
  { "distance_unit": "kilometers", "initial": 0.0 },
  { "time_unit": "minutes", "initial": 0.0 },
  { "energy_unit": "minutes", "initial": 0.0 }
]
```

but then when building the hashmap for the state model, it just uses the unit type as the feature name, so we get:

```rust
HashMap::from({
  ("distance", StateFeature::Distance(...)),
  ("time", StateFeature::Time(...)),
  ("energy", StateFeature::Energy(...)),
})
```

the keys for a PHEV should be "energy_electric" and "energy_liquid". StateModel deserialization should accept a JSON object with names as keys and StateFeatures as values, not as an array of StateFeatures.

### LINKED ISSUE #148: sort final state model configuration
a user may run multiple queries and find that the indices for their state features are never the same. to address this, we could enumerate the indices for the final set of features based on the lexicographical sorting of their feature names:

feature_name | current index | sorted index
--- | --- | ---
distance | 0 | 0
time | 1 | 2
energy | 2 | 1

as state representations expand in complexity, this would allow consistent deserialization for users inspecting their route state vectors, making it easier to perform route analysis.

edit: it would also be important to enforce downstream iteration over the features in index order during various serialization steps (route/tree state vector serialization, summary serialization) if it is not already the case.

### PR DESCRIPTION
**Pull request by robfitzgerald**
this PR addresses fixes to the StateModel, including:

- specializing the representation of state model for small instances (#138)
- deserializing from a JSON (#140)
- lexigraphical sorting of the representation (#148)

this increased the size of state_model.rs. to make reading easier, i pushed all internal ops into a pub(crate) trait for any methods that are only used internally. we're still looking at 649 lines.

i'm considering throwing out the UpdateOperation enum, since we're only using UpdateOperation::Replace, but i can imagine a world where we want them. if we want to further poke around at performance and want to reduce allocations, we may want to ease up on "replacing" and instead do floating point operations in-place.

Closes #138.
Closes #140.
Closes #148.

### CODE REVIEW

**Comment by nreinicke** [comment]
Cool stuff! I'm testing this out but getting some weird results:

<details>

<summary>Config</summary>

```toml
parallelism = 1
search_orientation = "edge"
response_persistence_policy = "persist_response_in_memory"
[response_output_policy]
type = "no_output"
# filename = "output.json"
# format = { type = "json", newline_delimited = true }

[graph]
edge_list_input_file = "../data/tomtom_denver/edges-compass.csv.gz"
vertex_list_input_file = "../data/tomtom_denver/vertices-compass.csv.gz"
verbose = true

[cost.vehicle_state_variable_rates.distance]
type = "factor"
factor = 0.655

# based on $20/hr approximation of 2023 median hourly wages, $/second
[cost.vehicle_state_variable_rates.time]
type = "factor"
factor = 0.333336

# based on AAA regular unleaded gas prices sampled 12/21/2023
[cost.vehicle_state_variable_rates.energy_liquid]
type = "factor"
factor = 3.120

# based on $0.50/kWh approximation of DCFC charge rates, $/kWhtype = "factor"
[cost.vehicle_state_variable_rates.energy_electric]
type = "factor"
factor = 0.50

# battery state shouldn't contribute to cost
[cost.vehicle_state_variable_rates.battery_state]
type = "factor"
factor = 0.0

[cost.state_variable_coefficients]
distance = 1
time = 1
energy_liquid = 1
energy_electric = 1
battery_state = 0 

[state]
distance = { distance_unit = "miles", initial = 0.0 }
time = { time_unit = "minutes", initial = 0.0 }
energy_liquid = { energy_unit = "gallons_gasoline", initial = 0.0 }
energy_electric = { energy_unit = "kilowatt_hours", initial = 0.0 }

[traversal]
type = "energy_model"
time_model_speed_unit = "kilometers_per_hour"
grade_table_input_file = "../data/tomtom_denver/edges-mean-grade-enumerated.txt.gz"
grade_table_grade_unit = "decimal"
output_time_unit = "minutes"
output_distance_unit = "miles"
# headings_table_input_file = "../data/tomtom_denver/edges-headings-enumerated.csv.gz"

[traversal.time_model]
type = "speed_table"
speed_table_input_file = "../data/tomtom_denver/speed-weekday-enumerated.txt.gz"
speed_unit = "kilometers_per_hour"
output_distance_unit = "miles"
output_time_unit = "minutes"

# [traversal.time_model]
# type = "time_of_day"
# edge_to_speed_profile_input_file = "../data/tomtom_denver/edges-speed-profile-mapping-enumerated.txt.gz"
# speed_profile_input_file = "../data/speed/speed-profile.csv.gz"
# # 35.704525623444134 is the average of all weekday_speed and weekend_speed entries
# fallback = { type = "global", fill_value = 35.704, speed_unit = "kilometers_per_hour" }
# max_speed = 128
# distance_unit = "miles"
# time_unit = "minutes"
# speed_unit = "miles_per_hour"

[traversal.time_model.speed_trace]
type = "dense"
speed_trace_input_file = "../data/speed/speed-trace.csv.gz"
stride = 300

[[traversal.vehicles]]
name = "2016_TOYOTA_Camry_4cyl_2WD"
type = "ice"
model_input_file = "../data/powertrain/2016_TOYOTA_Camry_4cyl_2WD.bin"
speed_unit = "miles_per_hour"
grade_unit = "decimal"
energy_rate_unit = "gallons_gasoline_per_mile"
ideal_energy_rate = 0.02857142857
real_world_energy_adjustment = 1.166
[traversal.vehicles.model_type.interpolate]
underlying_model_type = "smartcore"
speed_lower_bound = 0
speed_upper_bound = 100
speed_bins = 101
grade_lower_bound = -0.2
grade_upper_bound = 0.2
grade_bins = 41


[[traversal.vehicles]]
name = "2017_CHEVROLET_Bolt"
type = "bev"
model_input_file = "../data/powertrain/2017_CHEVROLET_Bolt.bin"
speed_unit = "miles_per_hour"
grade_unit = "decimal"
energy_rate_unit = "kilowatt_hours_per_mile"
battery_capacity = 60
battery_capacity_unit = "kilowatt_hours"
ideal_energy_rate = 0.2
real_world_energy_adjustment = 1.3958
[traversal.vehicles.model_type.interpolate]
underlying_model_type = "smartcore"
speed_lower_bound = 0
speed_upper_bound = 100
speed_bins = 101
grade_lower_bound = -0.2
grade_upper_bound = 0.2
grade_bins = 41

[frontier]
type = "combined"

[[frontier.models]]
type = "road_class"
road_class_input_file = "../data/tomtom_denver/edges-routing-class-enumerated.txt.gz"

[[frontier.models]]
type = "turn_restriction"
turn_restriction_input_file = "../data/tomtom_denver/edges-turn-restrictions.csv.gz"

[plugin]
input_plugins = [
    { type = "inject", key = "road_classes", value = "[1,2,3,4,5,6]", format = "json" },
    { type = "edge_rtree", geometry_input_file = "../data/tomtom_denver/edges-geometries-enumerated.txt.gz", road_class_input_file = "../data/tomtom_denver/edges-routing-class-enumerated.txt.gz" },
    { type = "grid_search" },
    { type = "load_balancer", weight_heuristic = { type = "haversine" } },
]
output_plugins = [
    { type = "summary" },
    { type = "traversal", route = "geo_json", tree = "geo_json", geometry_input_file = "../data/tomtom_denver/edges-geometries-enumerated.txt.gz" },
]
```

</details>

<details>

<summary>Query</summary>

```json
query = {
    "destination_y": 39.62627481432341,
    "destination_x": -104.99460207519721,
    "origin_y": 39.798311884359094,
    "origin_x": -104.86796368632217,
    "model_name": "2017_CHEVROLET_Bolt",
    "grid_search": {
        "test_cases": [
            {
                "name": "Least Distance",
                "state_variable_coefficients": {
                    "distance": 1,
                    "time": 0,
                    "energy_electric": 0,
                    "energy_liquid": 0,
                }
            },
            {
                "name": "Least Energy",
                "state_variable_coefficients": {
                    "distance": 0,
                    "time": 0,
                    "energy_electric": 1,
                    "energy_liquid": 0,
                }
            },
            {
                "name": "Least Time",
                "state_variable_coefficients": {
                    "distance": 0,
                    "time": 1,
                    "energy_electric": 0,
                    "energy_liquid": 0,
                }
            },
            {
                "name": "All",
                "state_variable_coefficients": {
                    "distance": 1,
                    "time": 1,
                    "energy_electric": 1,
                    "energy_liquid": 0,
                }
            },
        ]
    },
    "start_time": "09:00:00 AM",
    "start_weekday": "sunday",
}

```

</details>

Costs:

image

It looks like the same result is being returned regardless of the state variable coefficients. I also included "energy_liquid" in the state variable coefficient list since I got an error when I omitted it but it looks like there's some mismatch between the state and the vehicle since it should only be updating `energy_electric`

File: python/nrel/routee/compass/resources/osm_default_energy.toml [hunk 1]
```diff
@@ -1,6 +1,12 @@
 parallelism = 2
 search_orientation = "vertex"
 
+[state]
+distance = { distance_unit = "miles", initial = 0.0 }
+time = { time_unit = "minutes", initial = 0.0 }
+energy_liquid = { energy_unit = "gallons_gasoline_per_mile", initial = 0.0 }
```
Comment (nreinicke) [inline python/nrel/routee/compass/resources/osm_default_energy.toml:7]: I think the energy unit is "gallons_gasoline"

Comment (robfitzgerald) [reply python/nrel/routee/compass/resources/osm_default_energy.toml:7]: lol yup i know i mixed that up :-D

**Comment by nreinicke** [comment]
I'm also noticing that the traversal model doesn't have the final state in it, just the initial conditions:

```
'traversal_summary': {'battery_state': {'format': {'floating_point': {'initial': 100.0}},
   'name': 'soc',
   'unit': 'percent'},
  'distance': {'distance_unit': 'miles', 'initial': 0.0},
  'energy_electric': {'energy_unit': 'kilowatt_hours', 'initial': 0.0},
  'energy_liquid': {'energy_unit': 'gallons_gasoline', 'initial': 0.0},
  'time': {'initial': 0.0, 'time_unit': 'minutes'}},
```

**Comment by robfitzgerald** [comment]
> I'm also noticing that the traversal summary doesn't have the final state in it,

that's probably related to #141, i didn't dig into that here

**Comment by robfitzgerald** [comment]
> I'm testing this out but getting some weird results

those are weird results. i'm going to dive into this. i have a feeling that the CostModel::new method may be built on out-of-date assumptions about the state.

**Comment by nreinicke** [comment]
> that's probably related to #141, i didn't dig into that here

oh right! sorry, I missed that you hadn't included that

**Comment by robfitzgerald** [comment]
@nreinicke i think i've fixed the issues around how costmodel was tracking indices of things. i don't know if i love it; the cost model ditches the StateModel for it's own vectorized state tracking. but that's got to be faster than the StateModel. maybe that's best?

this might meet the needs of calling this PR "done", but as you pointed out, the summary is still getting overwritten by the state model:

```json
{
    "distance": {
        "distance_unit": "miles",
        "initial": 0.0
    },
    "energy_electric": {
        "energy_unit": "kilowatt_hours",
        "initial": 0.0
    },
    "energy_liquid": {
        "energy_unit": "gallons_gasoline",
        "initial": 0.0
    },
    "time": {
        "initial": 0.0,
        "time_unit": "minutes"
    }
}
```

**Comment by nreinicke** [comment]
Sweet, I just ran another test and it seems to be running fine when I supply this state section in the config:

```toml

[state]
distance = { distance_unit = "miles", initial = 0.0 }
time = { time_unit = "minutes", initial = 0.0 }
energy_electric = { energy_unit = "kilowatt_hours", initial = 0.0 }
```

But, if I add `energy_liquid`:

```toml

[state]
distance = { distance_unit = "miles", initial = 0.0 }
time = { time_unit = "minutes", initial = 0.0 }
energy_liquid = { energy_unit = "gallons_gasoline", initial = 0.0 }
energy_electric = { energy_unit = "kilowatt_hours", initial = 0.0 }
```

It uses `energy_liquid instead of time, here's one of the cost summaries:

```json
{
    "cost": {
        "distance": 26.493173210235792,
        "energy_electric": 9.788635799006627,
        "energy_liquid": 111.48210648921943,
        "time": 0.0,
        "total_cost": 147.76391549846187
    },
    "info": {
        "cost_aggregation": "sum",
        "network_state_variable_rates": [
            {
                "type": "zero"
            },
            {
                "type": "zero"
            },
            {
                "type": "zero"
            },
            {
                "type": "zero"
            }
        ],
        "state_variable_coefficients": [
            1.0,
            0.0,
            0.0,
            0.0
        ],
        "state_variable_indices": [
            [
                "distance",
                0
            ],
            [
                "energy_electric",
                1
            ],
            [
                "energy_liquid",
                2
            ],
            [
                "time",
                3
            ]
        ],
        "vehicle_state_variable_rates": [
            {
                "factor": 0.655,
                "type": "factor"
            },
            {
                "factor": 0.5,
                "type": "factor"
            },
            {
                "factor": 3.12,
                "type": "factor"
            },
            {
                "factor": 0.333336,
                "type": "factor"
            }
        ]
    }
}
```

Maybe this treads into the territory of how we want to improve the ergonomics of the state model configuration but we probably need to include both `energy_electric` and `energy_liquid` state defaults since we could be running either an electric or a conventional vehicle.

**Comment by robfitzgerald** [comment]
I think this is just me making a mistake again in the implementation.
There's this spot in the cost model new method where we build out vectors
of the data we need. I'm guessing that is still wrong, just because it's
hard to get right the way it was written before the state model existed. I
will take a look there and also will stub out the state model to see what
they both look like when instantiated.

**Comment by robfitzgerald** [comment]
@nreinicke i just rolled in the changes pushed to main from #150. does that fix the problem you're seeing with energy_electric / energy_liquid?

**Review by nreinicke — APPROVED** [review]
Yep, I just re-tested and the state indicies seem to be working now. Like you said, all the state variables are included but properly set to zero. I like the idea you proposed in #145 and that should address that particular problem

**Pull request merged**